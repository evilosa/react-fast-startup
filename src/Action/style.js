// @flow
export const ActionColors = {

};

export const createActionStyle = (colors = ActionColors) => ({
  colors,
});

export const ActionStyle = createActionStyle(ActionColors);

export default ActionStyle;