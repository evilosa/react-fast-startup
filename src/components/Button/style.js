import Color from 'color';

export const ButtonColors = {
  baseColor: '#fff',
  baseBackground: '#222222',
  primaryBackground: '#0074d9',
  warningBackground: '#f00',
};

export const createButtonStyle = (colors = ButtonColors) => ({
  base: {
    color: colors.baseColor,
    ':hover': {
      background: Color(colors.baseBackground).lighten(0.2).hex(),
    },
    background: colors.baseBackground,
  },

  primary: {
    background: colors.baseBackground,
  },

  small: {
    fontSize: '12px',
    padding: '5px',
    fontWeight: '300',
    borderRadius: '4px',
  },

  normal: {
    fontSize: '16px',
    fontWeight: '300',
    padding: '10px',
    borderRadius: '4px',
  },

  big: {
    fontSize: '24px',
    fontWeight: '300',
    padding: '15px',
    minWidth: '150px',
    borderRadius: '4px',
  },

  warning: {
    background: colors.warningBackground,
  }
});

export const ButtonStyle = createButtonStyle(ButtonColors);

export default ButtonStyle;
