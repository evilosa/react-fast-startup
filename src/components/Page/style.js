export const PageColors = {
  color: '#fff',
  background: '#222222',
};

export const createPageStyle = (/*colors = PageColors*/) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
    width: '100%',
    height: '100%',
  },

  header: {
    display: 'flex',
    flex: '1 0',
    height: 'auto',
  },

  content: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  footer: {
    display: 'flex',
    flex: '1 0',
    height: 'auto',
  }
});

export const ButtonStyle = createPageStyle();

export default ButtonStyle;
