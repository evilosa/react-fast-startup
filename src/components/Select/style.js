export const SelectColors = {
  color: '#fff',
  rootBorder: '#3b88ff',
  titleBorder: '#ff6f13',
  valueBorder: '#5bff42',
  optionsBorder: '#fcff19',
  searchBorder: '#ff0019',
  background: '#222222',
};

const defaultBorder = {
  borderWidth: '1px',
  borderStyle: 'solid',
}

export const createSelectStyle = (colors = SelectColors) => ({
  root: {
    display: 'flex',
    flex: '1 0',
    ...defaultBorder,
    borderColor: colors.rootBorder,
  },

  title: {
    display: 'flex',
    padding: '5px',
    alignItems: 'center',
    minWidth: '150px',
    ...defaultBorder,
    borderColor: colors.titleBorder,
  },

  value: {
    flex: '1 0',
    padding: '5px',
    ...defaultBorder,
    borderColor: colors.valueBorder,
  },

  optionsList: {
    display: 'block',
    position: 'absolute',
    backgroundColor: '#f9f9f9',
    maxHeight: '60px',
    minWidth: '160px',
    boxShadow: '0px 8px 16px 0px rgba(0,0,0,0.2)',
    marginTop: '30px',
    padding: '12px 16px',
    zIndex: '1',
    overflowY: 'auto',
    overflowX: 'hidden',
  },
});

export const SelectStyle = createSelectStyle();

export default SelectStyle;
