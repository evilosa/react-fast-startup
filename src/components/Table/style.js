export const TableColors = {
  baseBackground: '#fff',
  evenBackground: '#EEF3F5',
  border: '#D7DBDD',
  headerBackground: '#FCFDFD',
  headerText: '#16A2D7'
}

export const createTableStyle = (colors = TableColors) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  header: {
    display: 'flex',
    flex: '1 0',
    backgroundColor: colors.headerBackground,
    minHeight: '30px',
    maxHeight: '30px',
    borderBottomWidth: '1px',
    borderBottomStyle: 'solid',
    borderBottomColor: colors.border
  },

  body: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  footer: {
    display: 'flex',
    flex: '1 0',
    minHeight: '30px',
    maxHeight: '30px',
  },

  tableRow: {
    display: 'flex',
    flex: '1 0',
    background: colors.baseBackground,
    minHeight: '30px',
    maxHeight: '30px',
  },

  even: {
    background: colors.evenBackground,
  },

  tableHeaderColumn: {
    display: 'flex',
    flex: '1 0',
    padding: '5px',
    justifyContent: 'flex-start',
    alignItems: 'center',
    fontFamily: "'Raleway', sans-serif",
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '18px',
    fontStretch: '100%',
    color: colors.headerText,
  },

  tableRowActionColumn: {
    flex: '1 0',
    padding: '5px',
    fontFamily: "'Raleway', sans-serif",
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '18px',
    fontStretch: '100%',
  },

  tableRowColumn: {
    flex: '1 0',
    padding: '5px',
    fontFamily: "'Raleway', sans-serif",
    fontSize: '12px',
    fontWeight: '400',
    lineHeight: '18px',
    fontStretch: '100%',
  }
})

export const TableStyle = createTableStyle(TableColors)

export default TableStyle
