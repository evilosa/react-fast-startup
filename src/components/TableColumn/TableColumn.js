import React from 'react'
import PropTypes from 'prop-types'

const TableColumn = () => {
  return <div>Table column</div>
}

TableColumn.defaultProps = {
  header: '',
  propName: null,
  footer: null,
  width: '1 0',
  action: false,
}

TableColumn.propTypes = {
  header: PropTypes.string.isRequired,
  propName: PropTypes.string,
  width: PropTypes.string,
  action: PropTypes.bool,
  onClick: PropTypes.func,
}

TableColumn.displayName = 'TableColumn'

export default TableColumn