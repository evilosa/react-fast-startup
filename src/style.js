import '../fonts/Raleway/Raleway-Black.ttf'
import '../fonts/Raleway/Raleway-BlackItalic.ttf'
import '../fonts/Raleway/Raleway-Bold.ttf'
import '../fonts/Raleway/Raleway-BoldItalic.ttf'
import '../fonts/Raleway/Raleway-ExtraBold.ttf'
import '../fonts/Raleway/Raleway-ExtraBoldItalic.ttf'
import '../fonts/Raleway/Raleway-ExtraLight.ttf'
import '../fonts/Raleway/Raleway-ExtraLightItalic.ttf'
import '../fonts/Raleway/Raleway-Italic.ttf'
import '../fonts/Raleway/Raleway-Light.ttf'
import '../fonts/Raleway/Raleway-LightItalic.ttf'
import '../fonts/Raleway/Raleway-Medium.ttf'
import '../fonts/Raleway/Raleway-MediumItalic.ttf'
import '../fonts/Raleway/Raleway-Regular.ttf'
import '../fonts/Raleway/Raleway-SemiBold.ttf'
import '../fonts/Raleway/Raleway-SemiBoldItalic.ttf'
import '../fonts/Raleway/Raleway-Thin.ttf'
import '../fonts/Raleway/Raleway-ThinItalic.ttf'

// export const style = { black }