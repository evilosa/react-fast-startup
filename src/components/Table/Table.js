import * as React from 'react'
import Radium from 'radium'
import PropTypes from 'prop-types'
import style from './style'
import Button from '../Button'

class Table extends React.Component {

  static propTypes = {
    items: PropTypes.array.isRequired,
    canAddNewItem: PropTypes.bool,
    children: PropTypes.node,
    onAddItem: PropTypes.func,
  };

  static defaultProps = {
    items: [],
    canAddNewItem: false,
    header: '',
  };

  constructor(props) {
    super(props);

    const { items } = props;

    this.state = {
      items,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.items !== this.props.items) {
      this.setState(prev => ({
        ...prev,
        items: nextProps.items,
      }));
    }
  }

  _renderHeaders = (headerColumns) => {
    return headerColumns.map(({header, width, action}, index) => (
      <div className="table-header-column" key={index} style={[style.tableHeaderColumn, { flex: width }]}>{!action ? header : ''}</div>
    ));
  };

  _renderRows = (rowColumns) => {
    const { items } = this.state;
    return items.map((rowData, rowIndex) => (
      <div className="table-row" key={rowIndex} style={[style.tableRow, rowIndex % 2 === 0 && style.even]}>
        {this._renderRowColumns(rowColumns, rowData)}
      </div>
    ));
  };

  _renderRowColumns = (rowColumns, rowData) => (
    rowColumns.map((column, index) => {
      const columnData = rowData[column.propName];
      if (column.action) {
        return <div className="table-row-action" key={index} style={style.tableRowActionColumn}>{columnData}</div>
      }
      else {
        return <div className="table-row-column" key={index} style={style.tableRowColumn}>{columnData}</div>
      }
    })
  );

  render() {
    const { children, header, onAddItem } = this.props;
    const headerColumns = [];
    const rowColumns = [];
    const footerColumns = [];

    React.Children.forEach(children, (child) => {
      if (!child.type || child.type.displayName !== 'TableColumn')
        throw new Error('Child component should be instance of <TableColumn />');

      const { header, propName, footer, width, action } = child.props;

      // Fill headers columns
      headerColumns.push({ header, width, action });

      // Fill row columns
      rowColumns.push({ propName, width, action });

      // Fill footer
      footerColumns.push({ footer, width, action });
    });

    return (
      <div style={style.root}>
        <h1>{header}</h1>

        <div className="actions-pane-top">
          {onAddItem && <Button className="add-new-item-btn" onClick={onAddItem}>Add Item</Button>}
        </div>
        {/* Header */}
        <div
          className="table-header"
          style={style.header}
        >
          {this._renderHeaders(headerColumns)}
        </div>

        {/* Body */}
        <div
          className="table-body"
          style={style.body}
        >
          {rowColumns.length > 0 && this._renderRows(rowColumns)}
        </div>

        {/* Footer */}
        <div
          className="table-footer"
          style={style.footer}
        >
        </div>
      </div>
    );
  }
};

export default Radium(Table);