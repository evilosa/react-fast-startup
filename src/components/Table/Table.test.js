import React from 'react';
import { shallow, mount } from 'enzyme';
import Table from './Table';
import TableColumn from '../TableColumn/index';
import { createTableStyle } from './style'

describe('Table', () => {
  it('renders properly', () => {
    const table = shallow(<Table/>);
    expect(table).toMatchSnapshot();
  });

  it('should have correct style', () => {
    expect(createTableStyle()).toMatchSnapshot();
  })

  describe('props ', () => {
    const table = mount(<Table/>);
    const props = table.props();

    it('initialized with items', () => {
      expect(props.items).toEqual([]);
    });

    it('initialized with canAddNewItem option', () => {
      expect(props.canAddNewItem).toBe(false);
    });

    it('initialized with empty header', () => {
      expect(props.header).toEqual('');
    });
  });

  describe('state', () => {
    const table = mount(<Table/>);

    it('initialize default items state and update state with new items prop', () => {
      const testItem = { id: '1', name: 'Test name' };
      expect(table.state().items).toEqual([]);
      table.setProps({items: [ testItem ]});
      expect(table.state().items).toEqual([ testItem ]);
    });
  });

  describe('with empty items render', () => {
    const table = shallow(
      <Table>
        <TableColumn/>
        <TableColumn/>
        <TableColumn/>
      </Table>
    );

    it('headers', () => {
      expect(table.find('.table-header').exists()).toBe(true);
    });

    it('body', () => {
      expect(table.find('.table-body').exists()).toBe(true);
    });

    it('footer', () => {
      expect(table.find('.table-footer').exists()).toBe(true);
    });

    it('headers columns', () => {
      expect(table.find('.table-header-column').length).toEqual(3);
    });

    it('body without rows and row columns', () => {
      expect(table.find('.table-row').exists()).toBe(false);
      expect(table.find('.table-row-column').length).toEqual(0);
    });

    it('footer without columns', () => {
      expect(table.find('.table-footer-column').length).toEqual(0);
    });

    it('0 TableColumn with empty table', () => {
      expect(table.find(TableColumn).exists()).toBe(false);
    });
  });

  describe('with provided 2 items should renders correct', () => {
    const items = [
      { id: '1', title: 'Toast', sum: '$5'},
      { id: '2', title: 'Chicken', sum: '$15'},
    ];
    const table = shallow(
      <Table items={items}>
        <TableColumn/>
        <TableColumn/>
        <TableColumn/>
      </Table>
    );

    it('headers with 3 column', () => {
      expect(table.find('.table-header-column').length).toEqual(3);
    });

    it('body with 2 rows', () => {
      expect(table.find('.table-row').length).toEqual(2);
    });

    it('body with 6 row columns', () => {
      expect(table.find('.table-row-column').length).toEqual(6);
    });
  });

  describe('when contain TableColumn with action prop', () => {
    const items = [
      { id: '1', title: 'Toast', sum: '$5'},
      { id: '2', title: 'Chicken', sum: '$15'},
    ];

    const table = shallow(
      <Table items={items}>
        <TableColumn/>
        <TableColumn/>
        <TableColumn action/>
      </Table>
    );

    it('should render action', () => {
      expect(table.find('.table-row-action').length).toEqual(2);
    })
  })

  describe('with provided onAddItem', () => {
    const items = [
      { id: '1', title: 'Toast', sum: '$5'},
      { id: '2', title: 'Chicken', sum: '$15'},
    ]

    const mockAddItemFunction = jest.fn()

    const table = shallow(
      <Table
        items={items}
        onAddItem={mockAddItemFunction}
      >
        <TableColumn/>
        <TableColumn/>
        <TableColumn/>
      </Table>
    )

    it('should render Add item button', () => {
      expect(table.find('.add-new-item-btn').length).toEqual(1);
    })

    it('should call onAddItem function when add button clicked', () => {
      table.find('.add-new-item-btn').simulate('click')
      expect(mockAddItemFunction.mock.calls.length).toEqual(1)
    })
  })
})