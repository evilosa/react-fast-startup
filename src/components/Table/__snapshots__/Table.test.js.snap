// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders properly 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
      "flex": "1 0",
      "flexDirection": "column",
    }
  }
>
  <h1
    key=".0"
  />
  <div
    className="actions-pane-top"
    key=".1"
  />
  <div
    className="table-header"
    data-radium={true}
    key=".2"
    style={
      Object {
        "backgroundColor": "#FCFDFD",
        "borderBottomColor": "#D7DBDD",
        "borderBottomStyle": "solid",
        "borderBottomWidth": "1px",
        "display": "flex",
        "flex": "1 0",
        "maxHeight": "30px",
        "minHeight": "30px",
      }
    }
  />
  <div
    className="table-body"
    data-radium={true}
    key=".3"
    style={
      Object {
        "display": "flex",
        "flex": "1 0",
        "flexDirection": "column",
      }
    }
  />
  <div
    className="table-footer"
    data-radium={true}
    key=".4"
    style={
      Object {
        "display": "flex",
        "flex": "1 0",
        "maxHeight": "30px",
        "minHeight": "30px",
      }
    }
  />
</div>
`;
